implementation/building.tex

```diff
@@ -0,0 +1,165 @@
+\todo{Håvard: Jeg har ingen anelse over hvor riktig dette er. Noe av dette blir vel beskrevet et annet sted, vi får finne en måte å referere til riktig seksjon her, uten å gå altfor mye i detalj...}
+The flags were needed by \texttt{rustc} in order to tell it to link with an external library for a different architecture and to include a separate linker-script that took care of booting up the executable on this architecture.
+
+Another issue that was introduced by automatic compilation with Cargo, was how it structured the packages it compiled.
+When Cargo builds a project and its dependencies it structures all the generated metadata and the compiled libraries within a \texttt{target} directory, and an extra filename gets appended to all of these libraries.
+This extra filename is part of a hash that is generated based on the code in the library, and ensures that each and every build is consistent and it resolves any problems that might arise if several dependencies within a project depends on different versions of the same library.
+This works when Cargo handles the entire build process, but it our case, where we had to manually compile the final executable, it turned out to be a problem because the name of the library would change every time some of its content changed.
+We worked around this problem by modifying the build script to store the hash generated by Cargo for \texttt{emlib} to a text file, every time the library was built, and then included it in the makefile for the project.
+
+\subsection{Conditional linking with Cargo}
+
+The build process described in the previous section made it simpler to use third party libraries, but it did not solve all of our issues, the main problem that persisted was to have a good way of making \texttt{emlib} itself, portable.
+With the setup that we had, it was easy to create new executables \emph{within} the project, but it was hard to create new executables that \emph{depended} on \texttt{emlib}.
+Basically, because we had to work around Cargo in the final part of the build process, it also meant that \emph{every} project that wanted to depend on \texttt{emlib} also had implement the same workarounds.
+Thus, we needed to solve the problem of knowing where Cargo would store the project metadata, and we needed a way to get Cargo to compile the final executables with the extra linker-arguments needed by \texttt{rustc} in order to compile the binary for the Cortex-M3.
+
+Cargo does not have much documentation over how its internal works, or how to interfere with the build process, but the documentation does mention that if Cargo is to be invoked with a command that it does not have by default, it will query the system for this command.
+This means that if Cargo is invoked with e.g. the command \texttt{cargo foo <args>...}, it will query the system for an executable with the name \texttt{cargo-foo} and it will invoke this command with the trailing arguments if it exists.
```

Comment:
Is this just an arbitrary executable? In that case lets mention that this is the conventional way of exposing plugins for cargo.